### PR TITLE
build(mssql): add support for MSSQL 2022

### PR DESCRIPTION
**Pull request by WikiRik**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [X] Have you added new tests to prevent regressions?
- [X] Does `yarn test` or `yarn test-DIALECT` pass with this change (including linting)?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [X] Did you update the typescript typings accordingly (if applicable)?
- [X] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [X] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

Support for MSSQL 2022 has been added. MSSQL 2019 has not been removed yet since MSSQL 2022 is still in public preview; https://cloudblogs.microsoft.com/sqlserver/2022/05/24/announcing-sql-server-2022-public-preview-azure-enabled-with-continued-performance-and-security-innovation/

### Todos

- [x] Check website repo for updates to docs ( https://github.com/sequelize/website/pull/175 )
- [x] Remove CI check for MSSQL 2019 (in future PR)
- [x] Update required CI checks (in future PR)

### CODE REVIEW

**Review by sdepold — APPROVED** [review]
LGTM. We want to remove the mssql 2019 start scripts, right?

**Comment by WikiRik** [comment]
This already renames them, so we only have the 2022 scripts. But I'm not sure what we want to have in the `dev` folder, the oldest version we support or the newest one. What do you think?

**Comment by ephys** [comment]
Our MySQL scripts includes both v5 and v8. We could consider having two folders: `mssql/minimum` and `mssql/maximum` for each dialect

**Review by ephys — DISMISSED** [review]

**Comment by WikiRik** [comment]
That's also an option, but would we add (re)start/stop scrips for both as well? Like the yugabyte people proposed in their PR?

**Comment by ephys** [comment]
I think it would be a good idea

**Comment by WikiRik** [comment]
I'll add the scripts for 2017 here as well and then we should be good to merge

**Comment by WikiRik** [comment]
@mike-usa this is somewhat related to #14595 as well, I've limited this to just MSSQL but we can improve on this in the comments made by @ephys about the `start-all-minimum` and `start-all-maximum` in future PRs (or in yours).

**Comment by ephys** [comment]
Looking good :)

What do you think about naming the folders /oldest and /latest too? It would reduce the amount of work necessary to update these images (we could even add `/latest` to renovate)

**Comment by WikiRik** [comment]
That's a good idea! I'll see if I can tell renovate to ignore the `oldest` folder as well. We'll still have to update the docs and the CI workflow but at least we'll be informed by renovate of the new version

**Comment by ephys** [comment]
By the way I think renovate can be done in a separate PR

**Comment by WikiRik** [comment]
We'll do renovate later. But I think this is good to go then @ephys

File: dev/mssql/oldest/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+services:
+  mssql-2022:
```
Comment (ephys) [inline dev/mssql/oldest/docker-compose.yml:2]:
```suggestion
  mssql-oldest:
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -305,7 +305,7 @@ jobs:
       fail-fast: false
       matrix:
         node-version: [14, 18]
-        mssql-version: [2017, 2019]
+        mssql-version: [2017, 2019, 2022]
```
Comment (ephys) [inline .github/workflows/ci.yml:308]:
```suggestion
        mssql-version: [2017, 2022]
```

I think we should consider not including `2019` in our regular CI as it is already very long, until we optimize our tests to make them run faster and/or have 'intermediate' versions be run as smoke tests once a week?

File: dev/mssql/latest/docker-compose.yml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 services:
-  mssql-2019:
-    container_name: sequelize-mssql-2019
-    image: mcr.microsoft.com/mssql/server:2019-latest
+  mssql-2022:
```
Comment (ephys) [inline dev/mssql/latest/docker-compose.yml:2]:
```suggestion
  mssql-latest:
```

File: dev/mssql/latest/docker-compose.yml [hunk 4]
```diff
@@ -1,7 +1,7 @@
...
+    container_name: sequelize-mssql-2022
```
Comment (ephys) [inline dev/mssql/latest/docker-compose.yml:3]:
```suggestion
    container_name: sequelize-mssql-latest
```

File: dev/mssql/latest/docker-compose.yml [hunk 5]
```diff
@@ -15,4 +15,4 @@ services:
 
 networks:
   default:
-    name: sequelize-mssql-2019-network
+    name: sequelize-mssql-2022-network
```
Comment (ephys) [inline dev/mssql/latest/docker-compose.yml:18]: So other docker instances can access this network without needing to update :)

```suggestion
    name: sequelize-mssql-latest-network
```

File: dev/mssql/latest/start.sh [hunk 6]
```diff
@@ -3,14 +3,14 @@ set -Eeuxo pipefail # https://vaneyckt.io/posts/safer_bash_scripts_with_set_euxo
 cd -P -- "$(dirname -- "${BASH_SOURCE[0]}")" # https://stackoverflow.com/a/17744637
 
 
-docker-compose -p sequelize-mssql-2019 down --remove-orphans
-docker-compose -p sequelize-mssql-2019 up -d
+docker-compose -p sequelize-mssql-2022 down --remove-orphans
+docker-compose -p sequelize-mssql-2022 up -d
```
Comment (ephys) [inline dev/mssql/latest/start.sh:7]:
```suggestion
docker-compose -p sequelize-mssql-latest down --remove-orphans
docker-compose -p sequelize-mssql-latest up -d
```

File: dev/mssql/latest/start.sh [hunk 7]
```diff
@@ -3,14 +3,14 @@ set -Eeuxo pipefail # https://vaneyckt.io/posts/safer_bash_scripts_with_set_euxo
...
 
-./../../wait-until-healthy.sh sequelize-mssql-2019
+./../../wait-until-healthy.sh sequelize-mssql-2022
```
Comment (ephys) [inline dev/mssql/latest/start.sh:9]:
```suggestion
./../../wait-until-healthy.sh sequelize-mssql-latest
```

File: dev/mssql/latest/start.sh [hunk 8]
```diff
@@ -3,14 +3,14 @@ set -Eeuxo pipefail # https://vaneyckt.io/posts/safer_bash_scripts_with_set_euxo
...
 
-docker exec sequelize-mssql-2019 \
+docker exec sequelize-mssql-2022 \
```
Comment (ephys) [inline dev/mssql/latest/start.sh:11]:
```suggestion
docker exec sequelize-mssql-latest \
```

File: dev/mssql/oldest/docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+    container_name: sequelize-mssql-2017
```
Comment (ephys) [inline dev/mssql/oldest/docker-compose.yml:3]:
```suggestion
    container_name: sequelize-mssql-oldest
```

File: dev/mssql/oldest/docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+      - 22019:1433
+    healthcheck:
+      test: ["CMD", "/opt/mssql-tools/bin/sqlcmd", "-S", "localhost", "-U", "SA", "-P", "Password12!", "-l", "30", "-Q", "SELECT 1"]
+      interval: 3s
+      timeout: 1s
+      retries: 10
+
+networks:
+  default:
+    name: sequelize-mssql-2017-network
```
Comment (ephys) [inline dev/mssql/oldest/docker-compose.yml:18]:
```suggestion
    name: sequelize-mssql-oldest-network
```

File: dev/mssql/oldest/start.sh [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+set -Eeuxo pipefail # https://vaneyckt.io/posts/safer_bash_scripts_with_set_euxo_pipefail/
+cd -P -- "$(dirname -- "${BASH_SOURCE[0]}")" # https://stackoverflow.com/a/17744637
+
+
+docker-compose -p sequelize-mssql-2017 down --remove-orphans
+docker-compose -p sequelize-mssql-2017 up -d
+
+./../../wait-until-healthy.sh sequelize-mssql-2017
+
+docker exec sequelize-mssql-2017 \
```
Comment (ephys) [inline dev/mssql/oldest/start.sh:11]:
```suggestion
docker-compose -p sequelize-mssql-oldest down --remove-orphans
docker-compose -p sequelize-mssql-oldest up -d

./../../wait-until-healthy.sh sequelize-mssql-oldest

docker exec sequelize-mssql-oldest \
```

File: dev/mssql/oldest/stop.sh [hunk 12]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+set -Eeuxo pipefail # https://vaneyckt.io/posts/safer_bash_scripts_with_set_euxo_pipefail/
+cd -P -- "$(dirname -- "${BASH_SOURCE[0]}")" # https://stackoverflow.com/a/17744637
+
+
+docker-compose -p sequelize-mssql-2017 down --remove-orphans
```
Comment (ephys) [inline dev/mssql/oldest/stop.sh:6]:
```suggestion
docker-compose -p sequelize-mssql-oldest down --remove-orphans
```

File: package.json [hunk 13]
```diff
@@ -230,18 +230,21 @@
     "start-mysql": "bash dev/mysql/5.7/start.sh",
     "start-mysql-8": "bash dev/mysql/8.0/start.sh",
     "start-postgres": "bash dev/postgres/10/start.sh",
-    "start-mssql": "bash dev/mssql/2019/start.sh",
+    "start-mssql-oldest": "bash dev/mssql/2017/start.sh",
+    "start-mssql-latest": "bash dev/mssql/2022/start.sh",
```
Comment (ephys) [inline package.json:234]:
```suggestion
    "start-mssql-oldest": "bash dev/mssql/oldest/start.sh",
    "start-mssql-latest": "bash dev/mssql/latest/start.sh",
```

File: package.json [hunk 14]
```diff
@@ -230,18 +230,21 @@
...
     "start-db2": "bash dev/db2/11.5/start.sh",
     "stop-mariadb": "bash dev/mariadb/10.3/stop.sh",
     "stop-mysql": "bash dev/mysql/5.7/stop.sh",
     "stop-mysql-8": "bash dev/mysql/8.0/stop.sh",
     "stop-postgres": "bash dev/postgres/10/stop.sh",
-    "stop-mssql": "bash dev/mssql/2019/stop.sh",
+    "stop-mssql-oldest": "bash dev/mssql/2017/stop.sh",
+    "stop-mssql-latest": "bash dev/mssql/2022/stop.sh",
```
Comment (ephys) [inline package.json:241]:
```suggestion
    "stop-mssql-oldest": "bash dev/mssql/oldest/stop.sh",
    "stop-mssql-latest": "bash dev/mssql/latest/stop.sh",
```

**Review by ephys — CHANGES_REQUESTED** [review]
A few more places where we can remove the future need to update :)

File: package.json (hunk 13 above)
Comment (WikiRik) [reply package.json:234]: Woops, should have noticed these. It's still early 😅

File: .github/workflows/ci.yml (hunk 2 above)
Comment (WikiRik) [reply .github/workflows/ci.yml:308]: Yeah, I kept this in because 2022 is still in public preview but let's remove it now before we forget about it

Comment (WikiRik) [reply .github/workflows/ci.yml:308]: For running the 'intermediate' versions as smoke tests, I have plans on that which I still need to finish working out. But mainly the idea is indeed to run quite some additional configurations once a week

Comment (ephys) [reply .github/workflows/ci.yml:308]: Oh right I forgot you already told me

**Review by ephys — APPROVED** [review]
Running them locally went well

**Comment by mike-usa** [comment]
I really like this change!

I submit that the code should support more than 2 database versions. Supporting only two at the ends of the spectrum will catch 95% of bugs, but every version should be tested; at least a 3rd one somewhere between.  

At least if Sequelize says it supports from version X to version Y, it should test against those most recent major versions, nd include some minor versions.  Tests take too long as it is, so it's not feasible locally, but eventually... I think that should be a good vision to strive for.

**Comment by ephys** [comment]
> but every version should be tested; at least a 3rd one somewhere between.

That's planned, but not on every PR. Once a week is the current plan.

The probability that an intermediary version introduced a breaking change which is not detectable in the latest major is incredibly low. I think testing [oldest,latest] covers much more than 95% of issues. Oldest ensures we only use things that are available in all versions, latest ensures we're compatible with the breaking changes that happened since oldest.

**Pull request merged**